beep/protocol_tools/arbin_schedule_file.py

```diff
@@ -0,0 +1,251 @@
+        self.schedule_file_encoding = 'latin-1'
+
+    def hash_file(self, inputfile):
+        with open(inputfile, 'rb') as f:
+            chunk = f.read()
+        return hashlib.md5(chunk).digest()
+
+    def to_dict(self, inputfile):
+        """Schedule file ingestion. Converts a schedule file with section headers
+        to an ordered dict with section headers as nested dicts. One line in the
+        schedule file is not parsable by utf-8. This line is stored and returned
+        separately with the line number that it came from
+
+        Args:
+            inputfile (str): Schedule file (tested with FastCharge schedule file)
+
+        Returns:
+            sdu_dict (dict): Ordered dictionary with keys corresponding to options
+                or control variables. Section headers are nested dicts within the dict
+
+        """
+        sdu_dict = defaultdict(dict)
+        f = open(inputfile, 'rb')
+        lines = f.readlines()
+        keys = []
+        for line_num, line in enumerate(lines):
+            try:
+                line_plain = line.decode(self.schedule_file_encoding)
+            except UnicodeDecodeError:
+                print('Wrong encoding for schedule file at line: ' + str(line_num))
+            if re.search(self.section, line_plain) is not None:
+                if re.search(self.step, line_plain) is not None:
+                    if re.search(self.limit, line_plain) is not None:
+                        if len(keys) >= 3:
+                            keys = keys[:3]
+                            keys[2] = line_plain.strip('\r\n')
+                            sdu_dict[keys[0]][keys[1]][keys[2]] = OrderedDict({})
+                        else:
+                            keys.append(line_plain.strip('\r\n'))
+                            sdu_dict[keys[0]][keys[1]][keys[2]] = OrderedDict({})
+                    else:
+                        if len(keys) >= 2:
+                            keys = keys[:2]
+                            keys[1] = line_plain.strip('\r\n')
+                            sdu_dict[keys[0]][keys[1]] = OrderedDict({})
+                        else:
+                            keys.append(line_plain.strip('\r\n'))
+                            sdu_dict[keys[0]][keys[1]] = OrderedDict({})
+                else:
+                    if len(keys) >= 1:
+                        keys = keys[:1]
+                        keys[0] = line_plain.strip('\r\n')
+                    else:
+                        keys.append(line_plain.strip('\r\n'))
+            else:
+                key, value = line_plain.split('=', 1)
+                if len(keys) == 1:
+                    sdu_dict[keys[0]][key] = value.strip('\r\n')
+                if len(keys) == 2:
+                    sdu_dict[keys[0]][keys[1]][key] = value.strip('\r\n')
+                if len(keys) == 3:
+                    sdu_dict[keys[0]][keys[1]][keys[2]][key] = value.strip('\r\n')
+
+        sdu_dict = OrderedDict(sdu_dict)
```

Comment:
How important is the ordering? I don't think defaultdict guarantees that it will preserve the ordering, as read in from the file.

Reply:
I think the Joey's changes will make this an OrderedDict instead in order to guarantee the order of the objects.